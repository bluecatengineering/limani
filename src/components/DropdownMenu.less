// Copyright 2024 BlueCat Networks Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
@import '~@bluecateng/pelagos/less/elevations';
@import '~@bluecateng/pelagos/less/spacing';

.DropdownMenu {
    width: fit-content;

    &.Button--small {
        padding: 0 @sp-08;
    }

    &.Button--medium {
        padding: 0 @sp-12;
    }

    &.Button--large {
        padding: 0 @sp-16;
    }

    & > .Button__icon {
        margin: 0;
        transition: transform 0.15s ease-out;
    }

    &[aria-expanded='true'] > .Button__icon {
        margin: 0;
        transform: rotate(180deg);
    }

    &__popUp {
        @elv-04();
        position: absolute;
    }
}
