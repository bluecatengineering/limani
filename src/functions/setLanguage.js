/*
Copyright 2023 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import l10n from '../l10n';

const loaders = {
    en: () => import('../l10n/en.po'),
    zz: () => import('../l10n/zz.po'),
};

/**
 * Sets the language for Limani. The default language is English, there is no
 * need to call this function unless the selected language is different.
 * This function should be called as early as possible,
 * the caller must catch the returned promise and handle any errors.
 * @param {'en'|'es'} language The language translated content to load for.
 * @returns {*|Promise<never>}
 *
 * @example
 * import {setLanguage} from '@bluecateng/limani';
 *
 * setLanguage('es').catch((error) => {
 *     // handle error
 * })
 */
const setLanguage = (language) => {
    const loader = loaders[language];
    return loader
        ? loader().then(({ default: data }) => l10n.load(data))
        : Promise.resolve(console.warn(`Unknown language ${language}`));
};

export default setLanguage;
