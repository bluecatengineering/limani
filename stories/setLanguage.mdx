{/* Copyright 2023 BlueCat Networks Inc */}
{/*  */}
{/* Permission is hereby granted, free of charge, to any person obtaining a copy */}
{/* of this software and associated documentation files (the "Software"), to deal */}
{/* in the Software without restriction, including without limitation the rights */}
{/* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell */}
{/* copies of the Software, and to permit persons to whom the Software is */}
{/* furnished to do so, subject to the following conditions: */}
{/*  */}
{/* The above copyright notice and this permission notice shall be included in all */}
{/* copies or substantial portions of the Software. */}
{/*  */}
{/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR */}
{/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, */}
{/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE */}
{/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER */}
{/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, */}
{/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE */}
{/* SOFTWARE. */}
import { Meta } from '@storybook/addon-docs';

<Meta title='functions/setLanguage' />

# setLanguage

Sets the language for Limani. The default language is set to English i.e. `en`.
There is no need to call this function unless the selected
language is other than `en`. This function should be called as early as
possible, and wait before rendering the components,as it's an asynchronous operation.
Currently only `en` is supported.

## Parameters

<table>
    <thead>
        <tr>
            <th>Param</th>
            <th>Type</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>language</td>
            <td>`String`</td>
            <td>The language that needs to be set.</td>
        </tr>
    </tbody>
</table>

## Returns

`Promise` - Resolves successfully on language change.

## Example

```js
import { setLanguage } from '@bluecateng/limani';

const Example = () => (
    setLanguage('es').catch((error) => {
        // handle error
    })
    ...
);
```
