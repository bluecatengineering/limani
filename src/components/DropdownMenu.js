/*
Copyright 2024 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import {
    IconButton,
    Layer,
    Menu,
    MenuItem,
    MenuItemDivider,
    useLayer,
    useMenuHandler,
    useMenuPositioner,
} from '@bluecateng/pelagos';
import ChevronDown from '@carbon/icons-react/es/ChevronDown';
import PropTypes from 'prop-types';
import { Fragment, forwardRef, useEffect, useState } from 'react';
import { createPortal } from 'react-dom';
import { doGet } from '../functions/fetchFunctions';
import usePageMessages from './usePageMessages';
import usePageModalSpinner from './usePageModalSpinner';

import './DropdownMenu.less';

const setRefs =
    (...refs) =>
    (current) => {
        for (const ref of refs) {
            if (typeof ref === 'function') {
                ref(current);
            } else {
                ref.current = current;
            }
        }
    };

/** A button with a pop-up menu. */
const DropdownMenu = forwardRef(
    (
        {
            id,
            className,
            disabled,
            flipped,
            customValues,
            requestUrl,
            ...props
        },
        ref,
    ) => {
        const [defaultValue, setDefaultValue] = useState([]);
        const menuId = `${id}-menu`;
        const menuItems = customValues
            ? [...defaultValue, ...customValues]
            : defaultValue;
        const level = useLayer() + 1;
        const setPopUpPosition = useMenuPositioner(flipped);
        const { expanded, buttonProps, menuProps, buttonRef } =
            useMenuHandler(setPopUpPosition);
        const { setBusy } = usePageModalSpinner();
        const { addErrorMessage } = usePageMessages();

        useEffect(() => {
            setBusy(true);
            doGet(requestUrl || '/-/v3/dropdown_menu/data')
                .then((data) => {
                    const values = data?.dropdown_menu?.map((item) => ({
                        text: item?.text,
                        onClick: () => {
                            window.open(item?.url, '_blank').focus();
                        },
                    }));
                    setDefaultValue(values);
                })
                .catch((error) => addErrorMessage(error.message))
                .finally(() => setBusy(false));
        }, []);

        return (
            <>
                <IconButton
                    {...buttonProps}
                    {...props}
                    aria-controls={expanded ? menuId : null}
                    aria-expanded={expanded}
                    aria-haspopup='true'
                    className={`DropdownMenu${
                        className ? ` ${className}` : ''
                    }`}
                    data-layer={expanded ? level : null}
                    disabled={disabled || menuItems.length === 0}
                    icon={ChevronDown}
                    id={id}
                    ref={ref ? setRefs(ref, buttonRef) : buttonRef}
                />
                {expanded &&
                    createPortal(
                        <Layer className='DropdownMenu__popUp' level={level}>
                            <Menu
                                {...menuProps}
                                id={menuId}
                                aria-labelledby={id}>
                                {menuItems?.map((item, index) => (
                                    <Fragment key={index + item.text}>
                                        <MenuItem
                                            disable={item?.disable}
                                            onClick={item?.onClick}
                                            text={item.text}
                                        />
                                        {menuItems.length - 1 !== index && (
                                            <MenuItemDivider />
                                        )}
                                    </Fragment>
                                ))}
                            </Menu>
                        </Layer>,
                        document.body,
                    )}
            </>
        );
    },
);

DropdownMenu.displayName = 'DropdownMenu';

DropdownMenu.propTypes = {
    /** The component id. */
    id: PropTypes.string,
    /** The component class name(s). */
    className: PropTypes.string,
    /** The tooltip text to display. */
    tooltipText: PropTypes.string,
    /** The placement of the tooltip relative to the button. */
    tooltipPlacement: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
    /** The size of the button. */
    size: PropTypes.oneOf(['medium', 'large']),
    /** The button type. */
    type: PropTypes.oneOf(['primary', 'tertiary', 'ghost']),
    /** Whether the button is disabled. */
    disabled: PropTypes.bool,
    /** Whether the menu alignment should be flipped. */
    flipped: PropTypes.bool,
    // eslint-disable-next-line max-len
    /** Whether request called with a custom URL or use the default URL from Gateway at '/-/v3/dropdown_menu/data'*/
    requestUrl: PropTypes.string,
    /** The menu items. */
    customValues: PropTypes.arrayOf(
        PropTypes.shape({
            text: PropTypes.string.isRequired,
            onClick: PropTypes.func,
            disable: PropTypes.bool,
        }),
    ),
};

DropdownMenu.defaultProps = {
    size: 'medium',
    type: 'ghost',
};

export default DropdownMenu;
