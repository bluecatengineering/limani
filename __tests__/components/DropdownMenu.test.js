/*
Copyright 2024 BlueCat Networks Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { shallow } from 'enzyme';
import React from 'react';
import DropdownMenu from '../../src/components/DropdownMenu';
import setLanguage from '../../src/functions/setLanguage';

global.document = { body: 'body' };

jest.unmock('../../src/components/DropdownMenu');

const mockUsePageModalSpinner = {
    setBusy: jest.fn(),
};
jest.mock('../../src/components/usePageModalSpinner', () =>
    jest.fn(() => {
        return mockUsePageModalSpinner;
    }),
);

const mockAddErrorMessage = {
    addErrorMessage: jest.fn(),
};
jest.mock('../../src/components/usePageMessages', () =>
    jest.fn(() => {
        return mockAddErrorMessage;
    }),
);

const mockUseMenuHandler = {
    expanded: true,
    buttonProps: {
        onKeyDown: jest.fn(),
        onClick: jest.fn(),
    },
    menuProps: {
        onKeyDown: jest.fn(),
        onClick: jest.fn(),
    },
    guardProps: {},
};
jest.mock(
    '../../node_modules/@bluecateng/pelagos/dist/hooks/useMenuHandler',
    () =>
        jest.fn(() => {
            return mockUseMenuHandler;
        }),
);
const dropdownMenuMockDefaultValue = {
    /* eslint-disable camelcase */
    dropdown_menu: [
        {
            text: 'Go to API document',
            url: '/example_go_api_documents_by_help_links',
        },
    ],
};

const mockDoGet = {
    doGet: jest.fn(() =>
        Promise.resolve({ data: dropdownMenuMockDefaultValue }),
    ),
};
jest.mock('../../src/functions/fetchFunctions', () =>
    jest.fn(() => {
        return mockDoGet;
    }),
);

const dropdownMenuMockValue = [
    {
        text: 'Go to custom page',
        onClick: jest.fn(),
    },
];

describe('DropdownMenu', () => {
    describe('Rendering', () => {
        it('Render the DropdownMenu component default props', () => {
            const wrapper = shallow(<DropdownMenu />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('Render the DropdownMenu component with all props', () => {
            const wrapper = shallow(
                <DropdownMenu
                    id='test-id'
                    className='test_cn'
                    tooltipText='test-tooltip'
                    tooltipPlacement='left'
                    size='left'
                    type='tertiary'
                    flipped
                    customValues={dropdownMenuMockValue}
                />,
            );
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('Render the DropdownMenu component with pseudo translated text', () => {
            //Required for usePageMessages()
            setLanguage('zz').finally(() => {
                const wrapper = shallow(
                    <DropdownMenu
                        id='test-id'
                        className='test_cn'
                        tooltipText='test-tooltip'
                        tooltipPlacement='left'
                        size='left'
                        type='tertiary'
                        flipped
                        customValues={dropdownMenuMockValue}
                    />,
                );
                expect(wrapper.getElement()).toMatchSnapshot();
            });
        });
    });
});
