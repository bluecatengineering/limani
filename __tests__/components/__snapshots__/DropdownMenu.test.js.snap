// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu Rendering Render the DropdownMenu component default props 1`] = `
<React.Fragment>
  <IconButton
    aria-controls="undefined-menu"
    aria-expanded={true}
    aria-haspopup="true"
    className="DropdownMenu"
    data-layer={NaN}
    disabled={true}
    icon={[Function]}
    onClick={[MockFunction]}
    onKeyDown={[MockFunction]}
    size="medium"
    type="ghost"
  />
  <React.Portal
    container="body"
  >
    <Layer
      className="DropdownMenu__popUp"
      level={NaN}
    >
      <Menu
        id="undefined-menu"
        onClick={[MockFunction]}
        onKeyDown={[MockFunction]}
      />
    </Layer>
  </React.Portal>
</React.Fragment>
`;

exports[`DropdownMenu Rendering Render the DropdownMenu component with all props 1`] = `
<React.Fragment>
  <IconButton
    aria-controls="test-id-menu"
    aria-expanded={true}
    aria-haspopup="true"
    className="DropdownMenu test_cn"
    data-layer={NaN}
    disabled={false}
    icon={[Function]}
    id="test-id"
    onClick={[MockFunction]}
    onKeyDown={[MockFunction]}
    size="large"
    tooltipPlacement="left"
    tooltipText="test-tooltip"
    type="tertiary"
  />
  <React.Portal
    container="body"
  >
    <Layer
      className="DropdownMenu__popUp"
      level={NaN}
    >
      <Menu
        aria-labelledby="test-id"
        id="test-id-menu"
        onClick={[MockFunction]}
        onKeyDown={[MockFunction]}
      >
        <React.Fragment>
          <MenuItem
            onClick={[MockFunction]}
            text="Go to custom page"
          />
        </React.Fragment>
      </Menu>
    </Layer>
  </React.Portal>
</React.Fragment>
`;

exports[`DropdownMenu Rendering Render the DropdownMenu component with pseudo translated text 1`] = `
<React.Fragment>
  <IconButton
    aria-controls="test-id-menu"
    aria-expanded={true}
    aria-haspopup="true"
    className="DropdownMenu test_cn"
    data-layer={NaN}
    disabled={true}
    icon={[Function]}
    id="test-id"
    onClick={[MockFunction]}
    onKeyDown={[MockFunction]}
    size="large"
    tooltipPlacement="left"
    tooltipText="test-tooltip"
    type="tertiary"
  />
  <React.Portal
    container="body"
  >
    <Layer
      className="DropdownMenu__popUp"
      level={NaN}
    >
      <Menu
        aria-labelledby="test-id"
        id="test-id-menu"
        onClick={[MockFunction]}
        onKeyDown={[MockFunction]}
      />
    </Layer>
  </React.Portal>
</React.Fragment>
`;
